src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,551 @@
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGO AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, dateTimeProvider);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+        {
+            var relativeUri = Environment.GetEnvironmentVariable("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI");
+            var awsCredentialsCreators = new Func<AwsCredentials>[]
+            {
+                () => CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties),
+                () => CreateAwsCredentialsFromEnvironmentVariables(),
+                () => CreateAwsCredentialsFromEcsResponse(relativeUri),
+                () => CreateAwsCredentialsFromEc2Response()
+            };
+
+            foreach (var awsCredentialsCreator in awsCredentialsCreators)
+            {
+                var awsCredentials = awsCredentialsCreator();
+
+                if (awsCredentials != null)
+                {
+                    var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);
+                    return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, dateTimeProvider);
+                }
+            }
+
+            throw new ArgumentException("A MONGODB-AWS must have access key ID.");
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromMongoCredentials(string username, SecureString securePassword, IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            var sessionToken = ExtractSessionTokenFromMechanismProperties(properties);
+            ValidateCredentials(username, securePassword, sessionToken);
+
+            if (username == null)
+            {
+                return null;
+            }
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = securePassword,
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEnvironmentVariables()
+        {
+            var username = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
+            var password = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
+            var sessionToken = Environment.GetEnvironmentVariable("AWS_SESSION_TOKEN");
+            ValidateCredentials(username, password, sessionToken);
```

Comment:
I think this method should be responsible only for the creation of Aws credentials. The whole validation can be moved in one place where we iterate the created credentials. 
It will also allow us to avoid code duplication in the ValidateCredentials methods. The only reason why we have two methods now with fully the same idea is because we have passwords in two forms (string and SecurePassword), after this method, we will have only one variant of the password.

Reply:
Done.